### PR TITLE
[TEST]: Improvement: Paths manipulation: Regular flow

### PR DESCRIPTION
**Pull request by yuliiamir**
In the scope of this PR, the way of path manipulation was marginally modified(mostly using ISLs as a representation of the path instead of path nodes to get rid of PathHelper). The main modification includes a new representation of the available paths between switches for the cases with further path manipulation(make some path preferable, etc). By using a list of ISLs to represent a path, we avoid unnecessary converting during the test and get an easier way to compare objects.  Converting `List<List<PathNode>>` to  `List<Path> retrieveAvailablePaths` paths allows us to use all Path methods(getInvolvedSwitches()/getInvolvedIsls() that were a part of PathHelper class).

### CODE REVIEW

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/IslHelper.groovy [hunk 1]
```diff
@@ -120,11 +126,16 @@ class IslHelper {
 
     def setAvailableBandwidth(Isl isl, long newValue) {
         cleanupManager.addAction(RESET_ISL_PARAMETERS, {resetAvailableBandwidth([isl, isl.reversed])})
+        database.updateIslMaxBandwidth(isl, newValue)
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/IslHelper.groovy:129]: added this call as almost every test has both calls of updating maxBandwidth+availableBandwidth (it seems logically to update both of these values).

File: src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/IslHelper.groovy [hunk 2]
```diff
@@ -137,6 +148,21 @@ class IslHelper {
                 isl.dstPort, newMaxBandwidth)
     }
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/main/groovy/org/openkilda/functionaltests/helpers/IslHelper.groovy:150]: The rest of the new methods were migrated and slightly modified from PathHelper (we interact with the path by using Isls path representation and it seems obvious to move all these interactions to IslHelper).

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/BandwidthSpec.groovy [hunk 3]
```diff
@@ -187,9 +189,11 @@ class BandwidthSpec extends HealthCheckSpecification {
 
         // We need to handle the case when there are parallel links between chosen switches. So we make all parallel
         // links except the first link not preferable to avoid flow reroute when updating the flow.
-        List<List<PathNode>> parallelPaths = switchPair.paths.findAll { it.size() == 2 }
+        List<List<Isl>> parallelPaths = switchPair.retrieveAvailablePaths().collect { it.getInvolvedIsls() }
+                .findAll { it.size() == 1 }
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/BandwidthSpec.groovy:193]: 2 pathNodes == 1 Isls in path view

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy [hunk 4]
```diff
@@ -47,23 +45,19 @@ class IntentionalRerouteSpec extends HealthCheckSpecification {
...
-        alternativePaths.collect { altPath ->
-            def thinIsl = pathHelper.getInvolvedIsls(altPath).find {
-                !currentIsls.contains(it) && !currentIsls.contains(it.reversed)
-            }
-            def newBw = flow.maximumBandwidth - 1
+        def newBw = flow.maximumBandwidth - 1
+        alternativePathsIsls.collect { isls ->
+            def thinIsl = isls.find { !initialFlowIsls.contains(it) && !initialFlowIsls.contains(it.reversed) }
             [thinIsl, thinIsl.reversed].each {
-                database.updateIslMaxBandwidth(it, newBw)
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/IntentionalRerouteSpec.groovy:66]: removed this call as it is a part of  islHelper.setAvailableBandwidth

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy [hunk 5]
```diff
@@ -417,30 +417,27 @@ class FlowCrudSpec extends HealthCheckSpecification {
...
-        }.flatten() ?: assumeTrue(false, "No suiting active neighboring switches " +
-                "with two possible flow paths at least and different number of hops found")
-
-        and: "Make all shorter forward paths not preferable. Shorter reverse paths are still preferable"
-        List<Isl> modifiedIsls = possibleFlowPaths.findAll { it.size() == pathNodeCount }.collect {
-            pathHelper.getInvolvedIsls(it)
-        }.flatten().unique()
-        pathHelper.updateIslsCostInDatabase(modifiedIsls, Integer.MAX_VALUE)
+        List<List<Isl>> availablePathsIsls
+        int shortestIslCountPath = 1
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy:421]: 1 Isl == 2 pathNodes

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudV1Spec.groovy [hunk 6]
```diff
@@ -433,24 +432,20 @@ class FlowCrudV1Spec extends HealthCheckSpecification {
...
-            possibleFlowPaths = database.getPaths(it.srcSwitch.dpId, it.dstSwitch.dpId)*.path.sort { it.size() }
-            possibleFlowPaths.size() > 1 && possibleFlowPaths.max { it.size() }.size() > pathNodeCount
-        }.collect {
-            [it.srcSwitch, it.dstSwitch]
-        }.flatten() ?: assumeTrue(false, "No suiting active neighboring switches with two possible flow paths at least and " +
+        List<List<Isl>> availablePathsIsls
+        int shortestIslCountPath = 1
+        def swPair = switchPairs.all().neighbouring().withAtLeastNNonOverlappingPaths(2).getSwitchPairs().find {
+            availablePathsIsls = it.paths.collect { it.getInvolvedIsls() }
+            availablePathsIsls.size() > 1 && availablePathsIsls.find { it.size() > shortestIslCountPath}
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudV1Spec.groovy:439]: There is no need to find Path with max nodes count as 
`possibleFlowPaths.max { it.size() }.size() > pathNodeCount`  can be simplify to find any Path with more that expected nodes count (isl number) `availablePathsIsls.find { it.size() > shortestIslCountPath}`

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathCheckSpec.groovy [hunk 7]
```diff
@@ -143,44 +163,72 @@ class PathCheckSpec extends HealthCheckSpecification {
...
-            checkErrors.getValidationMessages().size() == 2
-            checkErrors.getValidationMessages().find { it.contains"The following segment intersects with the flow ${flow1.flowId}" }
-            checkErrors.getValidationMessages().find { it.contains"The following segment intersects with the flow ${flow2.flowId}" }
+            assert checkErrors.getValidationMessages().size() == 2
+            assert checkErrors.getValidationMessages().find { it.contains("The following segment intersects with the flow ${flow1.flowId}") }
+            assert checkErrors.getValidationMessages().find { it.contains("The following segment intersects with the flow ${flow2.flowId}") }
         }
     }
+
+    private static org.openkilda.messaging.payload.flow.FlowEncapsulationType convertEncapsulationType(FlowEncapsulationType origin) {
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathCheckSpec.groovy:206]: this method was moved from PathHelper

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowPingSpec.groovy [hunk 8]
```diff
@@ -157,26 +159,27 @@ class FlowPingSpec extends HealthCheckSpecification {
...
-        List<PathNode> aswitchPath
+        List<List<Isl>> allPaths = []
+        List<Isl> aswitchPathIsls = []
         //select src and dst switches that have an a-switch path
-        def (Switch srcSwitch, Switch dstSwitch) = [switches, switches].combinations()
-                .findAll { src, dst -> src != dst }.find { Switch src, Switch dst ->
-            allPaths = database.getPaths(src.dpId, dst.dpId)*.path
-            aswitchPath = allPaths.find { pathHelper.getInvolvedIsls(it).find { it.aswitch } }
-            aswitchPath
+        def swPair = switchPairs.all().getSwitchPairs().find {
```
Comment (yuliiamir) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowPingSpec.groovy:165]: switchPairs.all() - returns pairs with src != dst

**Review by IvanChupin — APPROVED** [review]
Great job!👍🏼

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathCheckSpec.groovy [hunk 9]
```diff
@@ -143,44 +163,72 @@ class PathCheckSpec extends HealthCheckSpecification {
...
+        // Let's laugh on this naive implementation after the third encapsulation type is introduced, not before.
```
[the remark this answers is not in the document]
Comment (yuliiamir) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathCheckSpec.groovy:207]: I moved this method from PathHelper and couldn't do it without this comment 😄

**Pull request merged**